service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,293 @@
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	replicationpb "go.temporal.io/server/api/replication/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/locks"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/service/history/hsm"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+const (
+	defaultPageSize = 32
+)
+
+type (
+	LastUpdatedStateTransitionGetter interface {
+		GetLastUpdateVersionedTransition() *persistencepb.VersionedTransition
+	}
+)
+
+func SyncWorkflowState(
+	ctx context.Context,
+	shardContext shard.Context,
+	request *historyservice.SyncWorkflowStateRequest,
+	workflowCache wcache.Cache,
+	logger log.Logger,
+) (_ *historyservice.SyncWorkflowStateResponse, retError error) {
+	wfCtx, releaseFunc, err := workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		shardContext,
+		namespace.ID(request.GetNamespaceId()),
+		request.Execution,
+		locks.PriorityLow,
+	)
+	defer func() {
+		if releaseFunc != nil {
+			releaseFunc(retError)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, shardContext)
+	if err != nil {
+		return nil, err
+	}
+	response := &historyservice.SyncWorkflowStateResponse{}
+	isSameBranch := workflow.TransitionHistoryStalenessCheck(mu.GetExecutionInfo().TransitionHistory, request.VersionedTransition) == nil
+	tombstoneBatch := mu.GetExecutionInfo().SubStateMachineTombstoneBatches
+	if isSameBranch &&
+		len(tombstoneBatch) > 0 && workflow.CompareVersionedTransition(tombstoneBatch[0].VersionedTransition, request.VersionedTransition) < 0 {
+		mutation, err := getMutation(mu, request.VersionedTransition)
+		if err != nil {
+			return nil, err
+		}
+		response.Attributes = &historyservice.SyncWorkflowStateResponse_Mutation{
+			Mutation: &replicationpb.SyncWorkflowStateMutationAttributes{
+				StateMutation:                     mutation,
+				ExclusiveStartVersionedTransition: request.VersionedTransition,
+			},
+		}
+	} else {
+		snapshot, err := getSnapshot(mu)
+		if err != nil {
+			return nil, err
+		}
+		response.Attributes = &historyservice.SyncWorkflowStateResponse_State{
+			State: snapshot,
+		}
+	}
+
+	newRunId := mu.GetExecutionInfo().NewExecutionRunId
+
+	releaseFunc(nil)
+	releaseFunc = nil
+
+	if len(newRunId) > 0 {
+		newRunInfo, err := getNewRunInfo(ctx, shardContext, workflowCache, namespace.ID(request.GetNamespaceId()), request.Execution.GetWorkflowId(), newRunId, logger)
+		if err != nil {
+			return nil, err
+		}
+		response.NewRunInfo = newRunInfo
+	}
+
+	events, err := getSyncStateEvents(ctx, shardContext, request.VersionHistories, mu.GetExecutionInfo().GetVersionHistories())
```

Comment:
we can't access `mu.GetExecutionInfo().GetVersionHistories()` after the lock is released